### ISSUE TITLE
InProcess: Sporadic 502 Bad Gateway responses to OPTION requests

### ISSUE THREAD

**Issue by Suchiman** [aspnet/AspNetCore#7021]
### Describe the bug
We're hosting an .NET Core framework-dependent ASP.NET Core 2.2 on App Services in Azure Germany with AspNetCoreModuleV2 as InProcess (which got recently enabled there). Sporadically we now notice that even in very low traffic situations, OPTION requests (from CORS) fail with 502 Bad Gateway. We did a load test of OPTION and POST requests to the same endpoint and while the OPTION sporadically failed with 502, the POST requests always worked. We've switched back to AspNetCoreModuleV2 using OutOfProcess and repeated the test and the issue went away. For sanity check we've reenabled InProcess and repeated the test which immediately yielded 502's again.

### To Reproduce
1. Create an ASP.NET Core 2.2 project using .NET Core and with this config
    ```csharp
    app.UseCors(builder => builder.AllowAnyHeader().AllowAnyMethod().AllowAnyOrigin().SetPreflightMaxAge(TimeSpan.FromMinutes(10)));
    ```
    and `SetCompatibilityVersion(CompatibilityVersion.Version_2_2);`
2. Hit this endpoint with a few hundred OPTION and POST requests, POST should always succeed while OPTION does not
    ```csharp
    [ApiController]
    [Route("api/[controller]")]
    [ResponseCache(NoStore = true, Location = ResponseCacheLocation.None)]
    public class IdController : Controller
    {
        [HttpPost("Ids")]
        public List<Guid> GenerateIds(int count = 200)
        {
            return Enumerable.Range(0, count).Select(_ => Guid.NewGuid()).OrderBy(x => new SqlGuid(x)).ToList();
        }
    }
    ```

### Expected behavior
Requests should always succeed

**Comment by Suchiman** [comment]
@jkotalik yes this is Azure App Service (see first sentence).
I'm using a Framework dependent deployment so i have to check on Azure App Service, right?

`dotnet --info` on Kudu gives
```
.NET Core SDK (reflecting any global.json):
 Version:   2.2.100
 Commit:    51868761f2

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.14393
 OS Platform: Windows
 RID:         win10-x86
 Base Path:   D:\Program Files (x86)\dotnet\sdk\2.2.100\

Host (useful for support):
  Version: 2.2.0
  Commit:  1249f08fed

.NET Core SDKs installed:
  1.1.10 [D:\Program Files (x86)\dotnet\sdk]
  2.1.402 [D:\Program Files (x86)\dotnet\sdk]
  2.1.500 [D:\Program Files (x86)\dotnet\sdk]
  2.2.100 [D:\Program Files (x86)\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.4 [D:\Program Files (x86)\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.6 [D:\Program Files (x86)\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.0 [D:\Program Files (x86)\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.4 [D:\Program Files (x86)\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.6 [D:\Program Files (x86)\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.0 [D:\Program Files (x86)\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 1.0.12 [D:\Program Files (x86)\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.9 [D:\Program Files (x86)\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [D:\Program Files (x86)\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.4 [D:\Program Files (x86)\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.6 [D:\Program Files (x86)\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.0 [D:\Program Files (x86)\dotnet\shared\Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```
which means they haven't patched to 2.2.1 yet it seems.

**Comment by jkotalik** [comment]
Try to do a standalone deployment with the ASP.NET 2.2.1: https://dotnet.microsoft.com/download/dotnet-core/2.2. That should resolve your issue.

**Comment by Suchiman** [comment]
@jkotalik indeed, that does fix it, thank you

**Comment by Suchiman** [comment]
@jkotalik just out of curiosity: 
> A 502 status code should not be returned from ANCM InProcess as it isn't a reverse proxy. Are you running this on Azure App Service or a similar hosting scenario which is returning 502?

That was my thought exactly, is there any reason why this still happened?

**Comment by jkotalik** [comment]
Azure web apps runs a load balancer in front of your application (see https://www.iis.net/downloads/microsoft/application-request-routing), so all failed requests to your app on Azure will result in 502s.

**Comment by sethberger** [comment]
I get sporadic 502s as well. My app is an InProcess, standalone 2.2.2 deployment. I am switching to OutOfProcess and hope the 502s go away! They cause serious business harm!

**Comment by gaborbuzasi** [comment]
This is still an issue and running a .net core 2.2 app on azure app service with the runtime of 2.2.6 this issue still comes up with the difference that it's not to OPTION requests but to actual GET requests.

**Comment by gaborbuzasi** [comment]
> 
> 
> Try to do a standalone deployment with the ASP.NET 2.2.1: https://dotnet.microsoft.com/download/dotnet-core/2.2. That should resolve your issue.

When you say trying to do a standalone deployment - what is exactly you mean in the context of an Azure App Service?

**Comment by Suchiman** [comment]
I'd assume a self contained deployment, see https://docs.microsoft.com/en-us/dotnet/core/deploying/#self-contained-deployments-scd

**Comment by gaborbuzasi** [comment]
Thanks @Suchiman, this seems to have fixed our issue with this too.
I will raise a separate issue that in framework dependent deployments with 2.2.6 runtime this issue seems to be coming up again.

**Comment by CarvellWakeman** [comment]
I'm not entirely sure how related our problem was. We're using InProcess hosting with AspNetCoreModuleV2 and runtime 2.2.6 (SDK 2.2.108). Our app is behind a load balancer. When hitting the load balancer **with InProcess hosting**, 204 NoContent responses were hanging. Seemed like the connection was left open by the server. 200 OK with no body did not have this issue.

With OutOfProcess hosting the issue did not occur. What we did to fix it was remove the Nuget package for Microsoft.AspNetCore (v2.2.0) and add Microsoft.AspNetCore.App with no version. Still unclear why this fixed it - 2.2.0 is the latest version of the AspNetCore package for Core 2.2.
